File: ruby/4_median_of_two_sorted_arrays.rb
# @param {Integer[]} nums1
# @param {Integer[]} nums2
# @return {Float}


def find_kth(nums1, nums2, k)
  m = nums1.length
  n = nums2.length

  return find_kth(nums2, nums1, k) if m > n # assume nums1 length less than nums2 length
  return nums2[k-1] if m == 0
  return [nums1[0], nums2[0]].min if k == 1

  i = [m, k/2].min
  j = [n, k/2].min

  if nums1[i-1] > nums2[j-1]
    return find_kth(nums1, nums2[j...n], k-j)
  else
    return find_kth(nums1[i...m], nums2, k-i)
  end
end


def find_median_sorted_arrays(nums1, nums2)
  m = (nums1.length + nums2.length + 1) / 2
  n = (nums1.length + nums2.length + 2) / 2

  return ( find_kth(nums1, nums2, m) + find_kth(nums1, nums2, n) ) / 2.0
end


puts find_median_sorted_arrays([1, 2], [3, 4])

File: cpp/4_median_of_two_sorted_arrays.cpp
#include <iostream>
#include <vector>
#include <string>
#include <stack>
#include <queue>
#include <map>
#include <set>
#include <unordered_set>
#include <unordered_map>
using namespace std;

// 给定两个有序数组，返回两个数组组合后的中位数。
// 可以化解成在两个有序数组中寻找第k个元素：
// 1. 当某个数组为空，直接返回[k-1];
// 2. 当k=1，返回min(nums1[0], nums2[0]);
// 3. 二分，减小k同时截取数组；
class Solution {
public:
	int findkth(vector<int> nums1, vector<int> nums2, int k) {
		if (nums1.empty()) { return nums2[k-1]; }
		if (nums2.empty()) { return nums1[k-1]; }
		if (k == 1) { return min(nums1[0], nums2[0]); }

	    int m = (int)nums1.size(), n = (int)nums2.size();
		int i = min(m, k/2), j = min(n, k/2);

		if (nums1[i-1] > nums2[j-1]) {
			return findkth(nums1, vector<int>(nums2.begin()+j, nums2.end()), k-j);
		} else {
			return findkth(vector<int>(nums1.begin()+i, nums1.end()), nums2, k-i);
		}
	}

    double findMedianSortedArrays(vector<int>& nums1, vector<int>& nums2) {
	    int m = (int)nums1.size(), n = (int)nums2.size();
		int i = (m+n+1) / 2, j = (m+n+2)/2;

		return (findkth(nums1, nums2, i) + findkth(nums1, nums2, j)) / 2.0;
    }
};

void unit_test() {
	Solution s;
}

int main() {
	unit_test();
	return 0;
}
